File: app/src/main/java/com/ocwvar/darkpurple/Services/MediaNotification.java
package com.ocwvar.darkpurple.Services;

import android.app.Notification;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Build;
import android.provider.MediaStore;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v7.app.NotificationCompat;
import android.text.TextUtils;

import com.ocwvar.darkpurple.Activities.SelectMusicActivity;
import com.ocwvar.darkpurple.AppConfigs;
import com.ocwvar.darkpurple.Bean.SongItem;
import com.ocwvar.darkpurple.R;
import com.ocwvar.darkpurple.Units.CoverImage2File;
import com.ocwvar.darkpurple.Units.Logger;

import java.io.IOException;

/**
 * Project DarkPurple
 * Created by 区成伟
 * On 2016/11/5 15:13
 * File Location com.ocwvar.darkpurple.Services
 * 多媒体Notification控制器
 */
class MediaNotification {

    static final int notificationID = 888;
    private static long notificationSwitchSkipTime = 0L;
    private final String TAG = MediaNotification.class.getSimpleName();
    private final Context context;
    private final MediaNotificationReceiver notificationReceiver;
    private MediaSessionCompat sessionCompat;

    MediaNotification(@NonNull Context context) {
        this.notificationReceiver = new MediaNotificationReceiver();
        this.context = context;
        setup();
    }

    /**
     * 获取更新Notification
     *
     * @param songItem    当前在播放序列的对象
     * @param audioStatus 当前的播放状态
     * @return 更新后的Notification
     */
    Notification updateNotification(@Nullable SongItem songItem, @NonNull AudioCore.AudioStatus audioStatus) {

        final NotificationCompat.Builder builder = new NotificationCompat.Builder(context);
        final NotificationCompat.MediaStyle mediaStyle = new NotificationCompat.MediaStyle(builder);

        mediaStyle.setShowActionsInCompactView(0, 1);

        mediaStyle.setMediaSession(sessionCompat.getSessionToken());

        //设置风格
        builder.setStyle(mediaStyle);

        //设置背景颜色
        builder.setColor(AppConfigs.Color.DefaultCoverColor);

        //不显示时间
        builder.setShowWhen(false);

        //仅通知一次
        builder.setOnlyAlertOnce(true);

        //设置点击时不取消
        builder.setAutoCancel(false);

        //设置状态栏图标
        builder.setSmallIcon(R.drawable.ic_action_small_icon);

        //设置锁屏是否显示
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            builder.setVisibility(Notification.VISIBILITY_PUBLIC);
        }

        //设置通知优先度
        builder.setPriority(Notification.PRIORITY_MAX);

        //设置通知类别为服务
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            builder.setCategory(Notification.CATEGORY_SERVICE);
        }

        //点击通知操作
        Intent intent = new Intent(context, SelectMusicActivity.class);
        intent.setAction(Intent.ACTION_MAIN);
        intent.addCategory(Intent.CATEGORY_LAUNCHER);
        PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, 0);
        builder.setContentIntent(pendingIntent);

        //根据播放状态不同 , 设置不同主按钮样式
        builder.addAction(generateAction(R.drawable.ic_media_previous, MediaNotificationReceiver.BUTTON_PREV));
        switch (audioStatus) {
            case Paused:
                builder.addAction(generateAction(R.drawable.ic_media_play, MediaNotificationReceiver.BUTTON_PLAY));
                break;
            case Playing:
                builder.addAction(generateAction(R.drawable.ic_media_pause, MediaNotificationReceiver.BUTTON_PAUSE));
                break;
        }
        builder.addAction(generateAction(R.drawable.ic_media_next, MediaNotificationReceiver.BUTTON_NEXT));

        //builder.addAction(generateAction(R.drawable.ic_media_close,MediaNotificationReceiver.BUTTON_CLOSE));

        Bitmap cover = loadCoverFromMainThread(songItem);
        if (cover == null) {
            cover = BitmapFactory.decodeResource(context.getResources(), R.drawable.ic_music_big);
        }

        if (songItem != null) {
            //更新标题
            builder.setContentTitle(songItem.getTitle());
            //更新作者
            builder.setContentText(songItem.getArtist());
            //更新封面
            builder.setLargeIcon(cover);
        } else {
            //更新标题
            builder.setContentTitle(context.getString(R.string.notification_string_empty));
            //更新作者
            builder.setContentText(context.getString(R.string.notification_string_empty));
            //封面设置为空
            builder.setLargeIcon(cover);
        }

        return builder.build();
    }

    /**
     * 生成按钮
     *
     * @param icon         图标资源
     * @param intentAction 按钮产生的广播Action
     * @return 按钮Action
     */
    private NotificationCompat.Action generateAction(int icon, String intentAction) {
        return new NotificationCompat.Action(icon, intentAction, PendingIntent.getBroadcast(context, 0, new Intent(intentAction), PendingIntent.FLAG_CANCEL_CURRENT));
    }

    /**
     * 初始化
     */
    private void setup() {
        sessionCompat = new MediaSessionCompat(this.context.getApplicationContext(), "test", new ComponentName(this.context.getApplicationContext().getPackageName(), HeadsetButtonReceiver.class.getName()), null);
        sessionCompat.setCallback(new MediaSessionCallback());
        sessionCompat.setFlags(MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS);
        sessionCompat.setActive(true);

        context.registerReceiver(notificationReceiver, notificationReceiver.filter);
    }

    /**
     * 结束资源
     */
    void close() {
        sessionCompat.setActive(false);
        sessionCompat.release();
        sessionCompat.setCallback(null);
        context.unregisterReceiver(notificationReceiver);
    }

    /**
     * 读取歌曲封面图像
     *
     * @param songItem 歌曲数据
     * @return 封面图像位图
     */
    private Bitmap loadCoverFromMainThread(@Nullable SongItem songItem) {

        final BitmapFactory.Options options = new BitmapFactory.Options();
        options.inPreferredConfig = Bitmap.Config.RGB_565;

        if (songItem == null) {
            return null;
        } else if (!TextUtils.isEmpty(songItem.getCustomCoverPath())) {
            return BitmapFactory.decodeFile(songItem.getCustomCoverPath(), options);
        } else if (songItem.getAlbumCoverUri() != null) {
            try {
                return MediaStore.Images.Media.getBitmap(context.getContentResolver(), songItem.getAlbumCoverUri());
            } catch (IOException e) {
                return null;
            }
        } else if (songItem.isHaveCover()) {
            return BitmapFactory.decodeFile(CoverImage2File.getInstance().getNormalCachePath(songItem.getPath()), options);
        } else {
            return null;
        }
    }

    /**
     * Notification 点击接收器
     */
    private class MediaNotificationReceiver extends BroadcastReceiver {

        static final String BUTTON_PLAY = "BUTTON_PLAY";
        static final String BUTTON_PAUSE = "BUTTON_PAUSE";
        static final String BUTTON_NEXT = "BUTTON_NEXT";
        static final String BUTTON_PREV = "BUTTON_PREV";
        static final String BUTTON_CLOSE = "BUTTON_CLOSE";

        public final IntentFilter filter;

        public MediaNotificationReceiver() {
            filter = new IntentFilter();
            filter.addAction(BUTTON_PLAY);
            filter.addAction(BUTTON_PAUSE);
            filter.addAction(BUTTON_NEXT);
            filter.addAction(BUTTON_PREV);
            filter.addAction(BUTTON_CLOSE);
        }

        @Override
        public void onReceive(Context context, Intent intent) {

            synchronized (this) {
                final long systemTime = System.currentTimeMillis();
                if (systemTime - notificationSwitchSkipTime < 500) {
                    return;
                } else {
                    notificationSwitchSkipTime = System.currentTimeMillis();
                }
            }

            final String action = intent.getAction();
            final Intent toService = new Intent();

            Logger.warnning(TAG, "接收到的事件: " + action);

            switch (action) {
                case BUTTON_PAUSE:
                    toService.setAction(AudioService.NOTIFICATION_PAUSE);
                    break;
                case BUTTON_PLAY:
                    toService.setAction(AudioService.NOTIFICATION_PLAY);
                    break;
                case BUTTON_NEXT:
                    toService.setAction(AudioService.NOTIFICATION_NEXT);
                    break;
                case BUTTON_PREV:
                    toService.setAction(AudioService.NOTIFICATION_PREV);
                    break;
                case BUTTON_CLOSE:
                    toService.setAction(AudioService.NOTIFICATION_CLOSE);
                    break;
            }

            context.sendBroadcast(toService);

        }

    }

    /**
     * MediaSession 回调 (主要用于接收耳机按钮回调)
     */
    private class MediaSessionCallback extends MediaSessionCompat.Callback {

        @Override
        public boolean onMediaButtonEvent(Intent mediaButtonEvent) {

            //接收到耳机事件之后就将事件转发到默认的接收器中

            final Intent intent = new Intent(HeadsetButtonReceiver.fromMediaSession);
            intent.putExtras(mediaButtonEvent.getExtras());
            context.sendBroadcast(intent);
            return true;
        }

    }

}

File: app/src/main/java/com/ocwvar/darkpurple/Services/MediaNotification.kt
package com.ocwvar.darkpurple.Services

import android.app.Notification
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.support.v7.app.NotificationCompat
import com.ocwvar.darkpurple.Activities.MainFrameworkActivity
import com.ocwvar.darkpurple.Bean.SongItem.SONGITEM_KEY_COVER_ID
import com.ocwvar.darkpurple.R
import com.ocwvar.darkpurple.Units.Cover.CoverManager

/**
 * Project DarkPurple
 * Created by OCWVAR
 * On 17-8-3 下午12:43
 * File Location com.ocwvar.darkpurple.Services
 * This file use to :   基于MediaSession的MediaStyle Notification
 */
class MediaNotification {

    val NOTIFICATION_ID: Int = 1451

    object ACTIONS {

        /**
         * Notification 动作 ：暂停
         */
        val NOTIFICATION_ACTION_PAUSE: String = "na1"

        /**
         * Notification 动作 ：播放
         */
        val NOTIFICATION_ACTION_PLAY: String = "na2"

        /**
         * Notification 动作 ：上一个媒体数据
         */
        val NOTIFICATION_ACTION_PREVIOUS: String = "na3"

        /**
         * Notification 动作 ：下一个媒体数据
         */
        val NOTIFICATION_ACTION_NEXT: String = "na4"

        /**
         * Notification 动作 ：关闭
         */
        val NOTIFICATION_ACTION_CLOSE: String = "na5"
    }

    fun createNotification(mediaSession: MediaSessionCompat, appContext: Context): Notification? {

        val controller: MediaControllerCompat? = mediaSession.controller
        val playbackState: PlaybackStateCompat? = controller?.playbackState
        val mediaMetadata: MediaMetadataCompat? = controller?.metadata

        if (controller == null || playbackState == null || mediaMetadata == null) {
            return null
        }

        //先检测当前的状态
        val state: Int = playbackState.state
        if (state == PlaybackStateCompat.STATE_ERROR || state == PlaybackStateCompat.STATE_NONE) {
            //不支持的状态
            return null
        }

        //创建Notification构造器
        val builder: NotificationCompat.Builder = NotificationCompat.Builder(appContext)
        builder
                //设置MediaSession
                .setStyle(NotificationCompat.MediaStyle()
                        //设置MediaSession的Token
                        .setMediaSession(mediaSession.sessionToken)
                        //设置显示的Action数量
                        .setShowActionsInCompactView(2)
                )
                //不显示Notification时间
                .setShowWhen(false)
                //状态栏上的图标
                .setSmallIcon(R.drawable.ic_action_small_icon)
                //设置最低隐私程度，允许在锁屏界面操作
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                //设置默认优先度
                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                //设置当Notification被移除时的广播（歌曲暂停时可以被移除）
                .setDeleteIntent(PendingIntent.getBroadcast(appContext, 0, Intent(ACTIONS.NOTIFICATION_ACTION_CLOSE), PendingIntent.FLAG_CANCEL_CURRENT))
                //开始显示时状态栏显示的文字： 艺术家 - 歌曲标题
                .setTicker(String.format("%s - %s", mediaMetadata.getString(MediaMetadataCompat.METADATA_KEY_ARTIST), mediaMetadata.getString(MediaMetadataCompat.METADATA_KEY_TITLE)))
                //主标题
                .setContentTitle(mediaMetadata.getString(MediaMetadataCompat.METADATA_KEY_TITLE))
                //副标题
                .setContentText(mediaMetadata.getString(MediaMetadataCompat.METADATA_KEY_ARTIST))
                //点击打开主界面
                .setContentIntent(PendingIntent.getActivity(appContext, 0, Intent(appContext, MainFrameworkActivity::class.java).let {
                    it.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                    it
                }, PendingIntent.FLAG_CANCEL_CURRENT))
                //添加 上一个 媒体数据按钮
                .addAction(android.R.drawable.ic_media_previous, "Previous", PendingIntent.getBroadcast(appContext, 0, Intent(ACTIONS.NOTIFICATION_ACTION_PREVIOUS), PendingIntent.FLAG_CANCEL_CURRENT))

        if (state == PlaybackStateCompat.STATE_PLAYING) {

            //当前是播放状态，添加 暂停播放 媒体按钮
            builder.addAction(android.R.drawable.ic_media_pause, "Pause", PendingIntent.getBroadcast(appContext, 0, Intent(ACTIONS.NOTIFICATION_ACTION_PAUSE), PendingIntent.FLAG_CANCEL_CURRENT))
        } else if (state == PlaybackStateCompat.STATE_PAUSED || state == PlaybackStateCompat.STATE_STOPPED) {

            //当前是暂停或停止状态，添加 开始播放 媒体按钮
            builder.addAction(android.R.drawable.ic_media_play, "Play", PendingIntent.getBroadcast(appContext, 0, Intent(ACTIONS.NOTIFICATION_ACTION_PLAY), PendingIntent.FLAG_CANCEL_CURRENT))
        }
        //添加 下一个 媒体数据按钮
        builder.addAction(android.R.drawable.ic_media_next, "Next", PendingIntent.getBroadcast(appContext, 0, Intent(ACTIONS.NOTIFICATION_ACTION_NEXT), PendingIntent.FLAG_CANCEL_CURRENT))

        //尝试获取有效封面图像
        BitmapFactory.decodeFile(CoverManager.getValidSource(mediaMetadata.getString(SONGITEM_KEY_COVER_ID)))?.let {
            builder.setLargeIcon(it)
        }


        return builder.build()
    }

}